drivers/at24cxxx/at24cxxx.c

```diff
@@ -0,0 +1,268 @@
+                  size_t len)
+{
+    int check = 0;
+    size_t wr = 0;
+    size_t chunk_len;
+
+    while (!check && (chunk_len = len - wr)) {
+        const uint8_t *chunk_data = ((const uint8_t *)data) + wr;
+        chunk_len =
+            min(chunk_len,
+                AT24CXXX_PAGE_SIZE - (pos & (AT24CXXX_PAGE_SIZE - 1)));
+        uint8_t polls = AT24CXXX_MAX_POLLS;
+        while ((check =
+                    i2c_write_regs(dev->params.i2c, dev->params.dev_addr, pos,
+                                   chunk_data,
+                                   chunk_len,
+                                   I2C_REG16)) == -ENXIO) {
+            if (--polls == 0) {
+                break;
+            }
+            xtimer_usleep(AT24CXXX_POLL_DELAY_US);
+        }
+        DEBUG("_write::i2c_write_regs check: %d; polls: %d\n", check, polls);
+        if (!check) {
+            wr += chunk_len;
+            pos += chunk_len;
+        }
+    }
+    return wr > 0 ? (int)wr : check;
+}
+
+static int _set(const at24cxxx_t *dev, uint32_t pos, uint8_t val, size_t len)
+{
+    int check = 0;
+    size_t wr = 0;
+    size_t chunk_len;
+
+    while (!check && (chunk_len = len - wr)) {
+        chunk_len =
+            min(chunk_len,
+                AT24CXXX_PAGE_SIZE - (pos & (AT24CXXX_PAGE_SIZE - 1)));
+        uint8_t polls = AT24CXXX_MAX_POLLS;
+        uint16_t reg = htons((uint16_t)pos);
+        while ((check =
+                    i2c_write_bytes(dev->params.i2c, dev->params.dev_addr, &reg,
+                                    AT24CXXX_ADDR_LEN,
+                                    I2C_NOSTOP)) == -ENXIO) {
+            if (--polls == 0) {
+                break;
+            }
+            xtimer_usleep(AT24CXXX_POLL_DELAY_US);
+        }
+        DEBUG("_set::i2c_write_bytes check: %d; polls: %d\n", check, polls);
+        for (size_t i = 0; !check && i < chunk_len; i++) {
+            check = i2c_write_byte(dev->params.i2c, dev->params.dev_addr, val,
+                                   i + 1 <
+                                   chunk_len ? I2C_NOSTART |
+                                   I2C_NOSTOP : I2C_NOSTART);
+            DEBUG("_set::i2c_write_byte check: %d\n", check);
+            if (!check) {
+                wr++;
+                pos++;
+            }
+        }
+    }
+    return wr > 0 ? (int)wr : check;
+}
+
+const at24cxxx_t *at24cxxx_get_default_dev(void)
+{
+    return (const at24cxxx_t *)&at24cxxx_params[0];
+}
+
+int at24cxxx_init(at24cxxx_t *dev, at24cxxx_params_t *params)
+{
+    if (!dev || !params) {
+        return -EINVAL;
+    }
+
+    dev->params = *params;
+    if (dev->params.pin_wp != GPIO_UNDEF) {
+        gpio_init(dev->params.pin_wp, GPIO_OUT);
+        at24cxxx_disable_write_protect(dev);
+    }
+    return AT24CXXX_OK;
+}
+
+uint8_t at24cxxx_read_byte(const at24cxxx_t *dev, uint32_t pos)
+{
+    assert(pos < dev->params.eeprom_size);
+
+    uint8_t val = 0;
+
+    i2c_acquire(dev->params.i2c);
```

Comment:
According to the doc of `i2c_acquire()`, it can fail. So the return value should be checked. The API does not allow to do any reasonable error handling.

I'm in favor for changing the API to have each function return success or an error code. But this means the API would no longer be compatible with the EEPROM API :-(. It seems indeed that e.g. the EEPROM used in ATmega CPUs cannot fail (or cannot report failures), so there was no use for an API being able to return error codes.

Reply:
There are implementations of `i2c_acquire()` that fail with -1 if the user enters a bad i2c bus index, some do an assert, some do not check the i2c index. Most drivers do not check the return value, so I think I just copied that. However it is of course better to check the return value, so I am going to change that.

You mean for the API of `...read_byte()` and `...write_byte()`? I did not like the API either, because the user will never notice that an error occurred. For the other functions, the user may notice that the number of bytes read or written does not match the number of bytes, he has specified.
That is why I set errno.

If you want me to change the API of the at24cxxx functions, I am going to do that.
But for the wrapper, changing the EEPROM API in general is again another issue to be handled in another PR, right?